rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py

```diff
@@ -0,0 +1,924 @@
+                -beta
+            )
+            B = 0
+            while B < batch_size_B:
+                # First, draw a random sample from Uniform(0, sum over all weights).
+                # Note, transitions with higher weight get sampled more often (as
+                # more random draws fall into larger intervals).
+                module_random_sum = (
+                    self.rng.random() * self._module_to_sum_segment[module_id].sum()
+                )
+                # Get the highest index in the sum-tree for which the sum is
+                # smaller or equal the random sum sample.
+                # Note, we sample `o_(t + n_step)` as this is the state that
+                # brought the information contained in the TD-error (see Schaul
+                # et al. (2018), Algorithm 1).
+                module_idx = self._module_to_sum_segment[module_id].find_prefixsum_idx(
+                    module_random_sum
+                )
+                # Get the theoretical probability mass for drawing this sample.
+                module_p_sample = (
+                    self._module_to_sum_segment[module_id][module_idx]
+                    / module_total_segment_sum
+                )
+                # Compute the importance sampling weight.
+                module_weight = (
+                    module_p_sample * self.get_num_timesteps(module_id)
+                ) ** (-beta)
+                # Now, get the transition stored at this index.
+                index_quadlet = self._module_to_indices[module_id][
+                    self._module_to_tree_idx_to_sample_idx[module_id][module_idx]
+                ]
+
+                # This will be an agent timestep (not env timestep).
+                # TODO (simon, sven): Maybe deprecate sa_episode_idx (_) in the index
+                #   quads. Is there any need for it?
+                ma_episode_idx, agent_id, sa_episode_ts = (
+                    index_quadlet[0] - self._num_episodes_evicted,
+                    index_quadlet[1],
+                    index_quadlet[2],
+                )
+
+                # Get the multi-agent episode.
+                ma_episode = self.episodes[ma_episode_idx]
+                # Retrieve the single-agent episode for filtering.
+                sa_episode = ma_episode.agent_episodes[agent_id]
+
+                # If we use random n-step sampling, draw the n-step for this item.
+                if random_n_step:
+                    actual_n_step = int(self.rng.integers(n_step[0], n_step[1]))
+                # If we cannnot make the n-step, we resample.
+                if sa_episode_ts + actual_n_step > len(sa_episode):
+                    continue
+                # Note, this will be the reward after executing action
+                # `a_(episode_ts)`. For `n_step>1` this will be the sum of
+                # all rewards that were collected over the last n steps.
+                sa_raw_rewards = sa_episode.get_rewards(
```

Comment:
perfect!

Reply:
I'm always worried about the DreamerV3 logic, which is different from all other algos wrt to sampling rewards from the buffer. The first reward in an episode is always 0.0 (it's basically a "reset" reward, similar to the "reset" zeroed-out-action).

Just need to be careful when we subclass the Dreamer's single-agent episode buffer.